src/clj/game/cards/ice.clj

```diff
@@ -78,6 +78,7 @@
          new-subs (->> (range total)
                        (reduce (fn [ice _] (add-sub ice sub (:cid ice) args)) new-card)
                        :subroutines
+                       (map-indexed (fn [idx sub] (assoc sub :index idx)))
```

Comment:
This shouldn't be necessary. It's already done by `add-ice`. Are you sure this is the source of the problem?

Reply:
Yes - the problem is when Envelopment reaches 0 counters, then `reset-variable-subs` gets called with `total` equal to 0 which means the `(range 0)` reduces with no calls to `add-ice`, so no subroutines get added but also the previous printed subroutines don't get their indexes reset.

If you remove this line you'll see the test fail with the only sub in Envelopment having an `:index` equal to `1` (instead of `0`).

Reply:
Ah yeah, range + reduce can be silly. This is a bit nit picky, but I'd prefer to only call map-indexed if total is 0. Add-sub encompasses everything needed to correctly add a sub, so it's not clear why this call is necessary.

Reply:
I wonder if like an `add-subs` function that takes a list of subs to add which then resets the indexes would be better or something.  Like the current way actually just recalculates the indexes n times which also seems pretty weird.  I wasn't sure all the places we are using `add-sub` where it's not in bulk already 🤔

Reply:
Yeah, recalculating the index n times is annoying but I couldn't figure out a way to only update the indexes from new-index onward (which is an easy thing to do in javascript lol).

Add-subs is hard because each sub might have a different index to be placed in which will influence the position of later subs, and subs are added one at a time according to the CR.